Port 2.x Annexes to 3.0 spec (#904)
spdx/spdx-spec: docs/annexes/including-security-information-in-SPDX.md

# 1. Including security information in a SPDX document

SPDX 3.0 has the concept of an [__External Reference__](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/ExternalRef.md) for an Element which points to "a resource outside the scope of the SPDX-3.0 content that provides additional characteristics of an Element." 

The specification for External Reference types has many [type options](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Vocabularies/ExternalRefType.md), a large handful of which pertain specifically to security use cases:

> * secureSoftwareAttestation: A reference to information assuring that the software is developed using security practices as defined by [NIST SP 800-218 Secure Software Development Framework (SSDF)](https://csrc.nist.gov/publications/detail/sp/800-218/final) or [CISA Secure Software Development Attestation Form](https://www.cisa.gov/sites/default/files/2023-04/secure-software-self-attestation_common-form_508.pdf).

Reviewer: NIST SP 800-218:
- Do we need specify the version number?
- "Secure Software Development Framework (SSDF) Version 1.1"

CISA Secure Software Development Attestation Form:
- The current link points to a draft version
- May need to update to the final version here: https://www.cisa.gov/resources-tools/resources/secure-software-development-attestation-form

Reviewer: I just realised that the text are from `ExternalRefType.md`, so I made PR for changes in the spdx-3-model source repo here as well: https://github.com/spdx/spdx-3-model/pull/708

Author: Thanks for the feedback @bact -- I will update my PR with your changes.

* securityAdvisory: A reference to a published security advisory (where advisory as defined per ISO 29147:2018) that may affect one or more elements, e.g., vendor advisories or specific NVD entries.
* securityAdversaryModel: A reference to the security adversary model for a package.
* securityFix: A reference to the patch or source code that fixes a vulnerability.
* securityOther: A reference to related security information of unspecified type.
* securityPenTestReport: A reference to a [penetration test](https://en.wikipedia.org/wiki/Penetration_test) report for a package.
* securityPolicy: A reference to instructions for reporting newly discovered security vulnerabilities for a package.
* securityThreatModel: A reference the [security threat model](https://en.wikipedia.org/wiki/Threat_model) for a package.

> * vulnerabilityDisclosureReport: A reference to a Vulnerability Disclosure Report (VDR) which provides the software supplier's analysis and findings describing the impact (or lack of impact) that reported vulnerabilities have on packages or products in the supplier's SBOM as defined in [NIST SP 800-161](https://csrc.nist.gov/publications/detail/sp/800-161/rev-1/final).

Reviewer: Update NIST SP 800-161 URL to https://csrc.nist.gov/pubs/sp/800/161/r1/final

Author: Updated.

The specification for External Reference types has many [type options](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Vocabularies/ExternalRefType.md), a large handful of which pertain specifically to security use cases:

* secureSoftwareAttestation
* securityAdvisory
* securityAdversaryModel
* securityFix
* securityOther
* securityPenTestReport
* securityPolicy
* securityThreatModel
* vulnerabilityDisclosureReport
* vulnerabilityExploitabilityAssessment

SPDX 3.0 also has the concept of [__External Identifier__](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/ExternalIdentifier.md) which should be used in cases where an identifier scheme exists and is already defined for an Element outside of SPDX-3.0.

There are several External Identifier [types](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Vocabularies/ExternalIdentifierType.md) that may be used in a security context:

* cpe22

> * cpe23

Author: Good catch, I think I forgot to update when you made this suggestion on my other PR.

* cve

> * cwe

Author: Good catch, I think I forgot to update when you made this suggestion on my other PR.

```json
{
  "@type": "Vulnerability",
  "@id": "urn:spdx.dev:cve-2020-2849",
  "summary": "Use of a Broken or Risky Cryptographic Algorithm",
  "description": "The npm package `elliptic` before version 6.5.4 are vulnerable to Cryptographic Issues via the secp256k1 implementation in elliptic/ec/key.js. There is no check to confirm that the public key point passed into the derive function actually exists on the secp256k1 curve. This results in the potential for the private key used in this implementation to be revealed after a number of ECDH operations are performed.",      
  "modified": "2021-03-08T16:06:43Z",
  "published": "2021-03-08T16:02:50Z",
  "externalIdentifiers": [
    {
      "@type": "ExternalIdentifier",
      "externalIdentifierType": "cve",
      "identifier": "CVE-2020-2849",
      "identifierLocator": [
        "https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-28498",
        "https://www.cve.org/CVERecord?id=CVE-2020-28498"
      ],
      "issuingAuthority": "urn:spdx.dev:agent-cve.org"
    },
    {
      "type": "ExternalIdentifier",

>       "externalIdentifierType": "securityOther",

Author: I took this from the vulnerability class example -- good catch. I will update the Vulnerability example as well.

To reference a CSAF Advisory document, include the document locator as an external reference of type `securityAdvisory` on a Package Element.

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-rh-open-shift",
  "name": "Red Hat OpenShift Enterprise",
  "packageVersion": "3.6",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityAdvisory",
      "locator": "https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/examples/csaf/rhsa-2019_1862.json"
    }
  ]
}
```

## G.1.3 Linking to a CycloneDX

To reference to [CycloneDX](https://cyclonedx.org) formatted security information applicable to a package you need to first create a Package Element. Then use an External Reference to link to the CycloneDX document which contains information about the package.

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-stack-cors",
  "name": "stack-cors",
  "packageVersion": "1.3.0",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityOther",
      "locator": "https://raw.githubusercontent.com/CycloneDX/bom-examples/ed522d1f051c364e045b87c20665003a0c4ea777/SBOM/laravel-7.12.0/bom.json"
    }
  ]
}

> ```

Reviewer 2: @goneall what do you think about this proposal to reference the cdx component URNs?

> ```

Reviewer 3: @puerco - agree with the URN approach - example LGTM

## G.1.4 Linking to an OSV

To include a reference to [Open Source Vulnerability](https://github.com/google/osv) (OSV) formatted security information applicable to a package you need to first create a Package Element. Then use an External Reference to link to the OSV advisory.

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-Django",
  "name": "Django",
  "packageVersion": "2.2",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityAdvisory",
      "locator": "https://github.com/github/advisory-database/blob/6b9d5bc96a62bb845ee71e4551a214eb1457e2c6/advisories/github-reviewed/2022/04/GHSA-2gwj-7jmv-h26r/GHSA-2gwj-7jmv-h26r.json"
    }
  ]
}
```

> ## G.1.5 Linking to an OmniBOR (formerly known as GitBOM)
> 
> To reference an [OmniBOR](https://omnibor.io/) (Universal Bill Of Receipts) formatted security information applicable to a package you must first create a Package Element. Then use an External Identifier to link to the OmniBOR document.
> **TODO: Is this gitoid correct? do we need a different example here?**

Reviewer 2: Gitoids identify the package blob, they are an identifier not a reference to something else.

Regarding the TODO, yes it is correct. The string in the example is the IANA URL scheme for gitoids.

```suggestion
## G.1.5 Identifying a Package with an OmniBOR Gitoid 

To identify a Package with an [OmniBOR](https://omnibor.io/) (Universal Bill Of Receipts, formerly known as GitBOM) gitoid, use an External Identifier to add gitoid to the package.
```

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-example",
  "name": "Example",
  "packageVersion": "1.2.3",
  "externalIdentifiers": [
    {
      "@type": "ExternalIdentifier",
      "externalIdentifierType": "gitoid",
      "identifier": "gitoid:blob:sha1:bcb99b819dadaebdf2c8f88d92ee9024c45f9df3"
    }
  ]
}
```

> ## G.1.6 Linking to a vulnerability disclosure document

Reviewer 2: What is the difference between this and an advisory? (Especially since the links point to advisory data)

    }
  ]
}
```

To communicate that a package is not vulnerable to a specific vulnerability it is recommended to reference a web page indicating why given vulnerabilities are not applicable using an External Reference on the package.

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:example-1",
  "name": "example",
  "packageVersion": "1.0.0",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityAdvisory",
      "locator": "https://example.com/product-x/security-info-not-affected.html"
    }
  ]
}
```

To refer to a security disclosure feed, such as the security bulletins from [CERT-EU](https://cert.europa.eu), include an External Reference in the package Element.

**TODO: We need a new link as current one is invalid**

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:example-2",
  "name": "example",
  "packageVersion": "2.0.0",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityAdvisory",
      "locator": "https://cert.europa.eu/cert/Data/newsletter/reviewlatest-SecurityBulletins.xml"
    }
  ]
}
```

> ## G.1.7 Linking to a code fix for a security issue

Reviewer 2: ```suggestion
## G.1.7 Linking to a Code Fix for a Security Issue
```

```json
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityFix",
      "locator": "https://github.com/indutny/elliptic/commit/441b7428b0e8f6636c42118ad2aaa186d3c34c3f",
      "comment": "elliptic before version 6.5.4 are vulnerable to Cryptographic Issues via the secp256k1 implementation in elliptic/ec/key.js. This patch fixes CVE-2020-28498."
    }
  ]
```

A fix reference may point to a configuration change for example the patch file as one of the fixes for [CVE-2022-26499](https://nvd.nist.gov/vuln/detail/CVE-2022-26499).

```json
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityFix",
      "locator": "https://downloads.digium.com/pub/security/AST-2022-002-16.diff"
    }
  ]
```

Alternatively, it may also link to a landing page with patches for a variety of products such as Oracle patch information for [CVE-2021-44228](https://nvd.nist.gov/vuln/detail/CVE-2021-44228).

```json
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityFix",
      "locator": "https://www.oracle.com/security-alerts/cpujan2022.html"
    }
  ]
```

> ## G.1.8 Linking to any security related document

Reviewer 2: ```suggestion
## G.1.8 Linking to Any Security Related Document
```

If you want to reference any security information related to a package but cannot or do not wish to specify its kind, use the `securityOther` externalRefType.

```json
{
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-elliptic",
  "name": "elliptic",
  "packageVersion": "6.5.4",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityOther",
      "locator": "https://github.com/christianlundkvist/blog/blob/aa3a69b5e4c06e4435070610c0c4a2b1e8731783/2020_05_26_secp256k1_twist_attacks/secp256k1_twist_attacks.md",
      "comment": "Blog post from author who wrote fix for CVE-2020-28498."
    }
  ]
}
```

One can also use it to refer to guidance related to a vulnerability such as CISA guidance for Apache Log4j.

```json
  "@type": "Package",
  "@id": "urn:spdx.dev:pkg-apache-log4j",
  "name": "log4j",
  "packageVersion": "2.14.0",
  "externalRefs": [
    {
      "@type": "ExternalRef",
      "externalRefType": "securityOther",
      "locator": "https://www.cisa.gov/uscert/apache-log4j-vulnerability-guidance"
    }
  ]
```

> ## G.1.9 Linking to an SBOM vulnerability report for a Software Product (per NIST Executive Order 14028)

Reviewer 2: Isn't this simply a VDR?

```suggestion
## G.1.9 Linking to a Vulnerability Disclosure Report (VDR)
```
